pretyping/reductionops.ml

```diff
@@ -1423,23 +1426,43 @@ let is_arity env sigma c =
 (*************************************)
 (* Metas *)
 
-let meta_value env evd mv =
-  let rec valrec mv =
-    match meta_opt_fvalue evd mv with
-    | Some (b,_) ->
-      let metas = Metamap.bind valrec b.freemetas in
-      instance env evd metas b.rebus
-    | None -> mkMeta mv
+type meta_instance_subst = {
+  sigma : Evd.evar_map;
+  mutable cache : EConstr.t Metamap.t;
+}
+
+let create_meta_instance_subst sigma = {
+  sigma;
+  cache = Metamap.empty;
+}
+
+let eval_subst env subst =
+  let self = ref (fun _ -> assert false) in
+  let ans mv =
```

Comment:
Why use a ref instead of `let rec ans mv = ...`?

Reply:
It's remnant of an old code where this was prohibited for syntactic reasons. Will fix.